### PR TITLE
Add SameSite setting for cookies

### LINKED ISSUE #5583: SameSite Setting for Cookies
- Gitea version (or commit ref): 1.6.2
- Git version: 2.20.1
- Operating system: CentOS 7
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [x] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [x] Yes (provide example URL) [https://observatory.mozilla.org/analyze/try.gitea.io](https://observatory.mozilla.org/analyze/try.gitea.io), see "Anti-CSRF tokens set without using the SameSite flag"
  - [ ] No
  - [ ] Not relevant
- Log gist: N/A

## Description
The [SameSite](https://www.owasp.org/index.php/SameSite) setting should be enabled on the session and CSRF cookies as an added prevention against CSRF. [Mozilla](https://blog.mozilla.org/security/2018/04/24/same-site-cookies-in-firefox-60/) does a good job of explaining its purpose, but the gist is that it prevents cookies being sent in a request initiated from a foreign origin.

## Screenshots
N/A

### PR DESCRIPTION
**Pull request by zeripath**
Add SameSite setting for cookies and rationalise the cookie setting code. Switches SameSite to Lax by default. 

There is a possible future extension of differentiating which cookies could be set at Strict by default but that is for a future PR.

Fix #5583

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

**Comment by zeripath** [comment]
Please check and test I've set these correctly - I am by no means certain I've set these right.

There are also multiple places where we essentially have the same code duplicated multiple times so we should really do a small refactor sorting this out.

I've also taken the liberty of removing the crazy 1<<31-1 time from the lang attribute from the bottom menu.

**Comment by zeripath** [comment]
I've marked this as 1.14 as I suspect that between 1.14 being released and 1.15 being released we may find ourselves in a situation whereby this is forced upon us.

Better to put it in now than later.

**Comment by 6543** [comment]
@zeripath `panic: session.Options.SameSite: readUint64: unexpected character: , error found in #10 byte of ...|ameSite":"strict"}|..., bigger context ...|ime":86400,"Secure":false,"Domain":"","SameSite":"strict"}|...
`

File: docs/content/doc/advanced/config-cheat-sheet.en-us.md [hunk 1]
```diff
@@ -557,6 +557,8 @@ Define allowed algorithms and their minimum key length (use -1 to disable a type
 - `COOKIE_NAME`: **i\_like\_gitea**: The name of the cookie used for the session ID.
 - `GC_INTERVAL_TIME`: **86400**: GC interval in seconds.
 - `SESSION_LIFE_TIME`: **86400**: Session life time in seconds, default is 86400 (1 day)
+- `DOMAIN`: **\<empty\>**: Sets the cookie Domain
```
Comment (lafriks) [inline docs/content/doc/advanced/config-cheat-sheet.en-us.md:560]: By default imho it should set http domain imho

**Comment by lafriks** [comment]
It should also set path to cookies

**Review by 6543 — APPROVED** [review]

File: modules/web/middleware/locale.go [hunk 2]
```diff
@@ -42,8 +43,30 @@ func Locale(resp http.ResponseWriter, req *http.Request) translation.Locale {
 	}
 
 	if changeLang {
-		SetCookie(resp, "lang", lang, 1<<31-1)
+		SetLocaleCookie(resp, lang)
```
Comment (zeripath) [inline modules/web/middleware/locale.go]: This is a potential significant change we need to be sure I'm correct here.

File: docs/content/doc/advanced/config-cheat-sheet.en-us.md (hunk 1 above)
Comment (zeripath) [reply docs/content/doc/advanced/config-cheat-sheet.en-us.md:560]: As per discussion on discord. See https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie .
Leaving this blank is actually the correct thing to do.

File: modules/web/middleware/locale.go (hunk 2 above)
Comment (6543) [reply modules/web/middleware/locale.go]: What was this `1<<31-1` for?

Comment (zeripath) [reply modules/web/middleware/locale.go]: So previously clicking on the language selector at the bottom of the screen would set a permanent lang cookie for unlogged in users ensuring that they always got their correct language. 

It's kinda a bad thing to set a permanent cookie and there moans about it in one of the issues.

TBH if you're not logged in you shouldn't necessarily get to have permanent settings but I'm not sure.

Comment (lafriks) [reply modules/web/middleware/locale.go]: If you have selected language why would you not want to keep it? This way login will be pretty much always in not the language you selected

Comment (zeripath) [reply modules/web/middleware/locale.go]: I'll put it back in and we should look at this another time.

**Review by lafriks — APPROVED** [review]

**Comment by zeripath** [comment]
@CirnoT I recall you had thoughts on defaulting to strict and that we should be defaulting to lax?

**Comment by zeripath** [comment]
Just need to ensure that strict is the right default I have a feeling we should have lax actually

**Comment by zeripath** [comment]
Yup we should default to lax

File: custom/conf/app.example.ini [hunk 3]
```diff
@@ -794,6 +794,8 @@ COOKIE_SECURE = false
 GC_INTERVAL_TIME = 86400
 ; Session life time in seconds, default is 86400 (1 day)
 SESSION_LIFE_TIME = 86400
+; SameSite settings. Either "none", "lax", or "strict"
+SAME_SITE=strict
```
Comment (zeripath) [inline custom/conf/app.example.ini]:
```suggestion
SAME_SITE=lax
```

File: docs/content/doc/advanced/config-cheat-sheet.en-us.md [hunk 4]
```diff
@@ -557,6 +557,8 @@ Define allowed algorithms and their minimum key length (use -1 to disable a type
...
+- `SAME_SITE`: **strict** \[strict, lax, none\]: Set the SameSite setting for the cookie.
```
Comment (zeripath) [inline docs/content/doc/advanced/config-cheat-sheet.en-us.md]:
```suggestion
- `SAME_SITE`: **lax** \[strict, lax, none\]: Set the SameSite setting for the cookie.
```

File: modules/setting/session.go [hunk 5]
```diff
@@ -52,6 +56,15 @@ func newSessionService() {
 	SessionConfig.Gclifetime = sec.Key("GC_INTERVAL_TIME").MustInt64(86400)
 	SessionConfig.Maxlifetime = sec.Key("SESSION_LIFE_TIME").MustInt64(86400)
 	SessionConfig.Domain = sec.Key("DOMAIN").String()
+	samesiteString := sec.Key("SAME_SITE").In("strict", []string{"none", "lax", "strict"})
+	switch strings.ToLower(samesiteString) {
+	case "none":
+		SessionConfig.SameSite = http.SameSiteNoneMode
+	case "lax":
```
Comment (zeripath) [inline modules/setting/session.go]:
```suggestion
	case "strict":
```

File: modules/setting/session.go [hunk 6]
```diff
@@ -52,6 +56,15 @@ func newSessionService() {
...
+		SessionConfig.SameSite = http.SameSiteLaxMode
```
Comment (zeripath) [inline modules/setting/session.go]:
```suggestion
		SessionConfig.SameSite = http.SameSiteStrictMode
```

File: modules/setting/session.go [hunk 7]
```diff
@@ -52,6 +56,15 @@ func newSessionService() {
...
+	default:
+		SessionConfig.SameSite = http.SameSiteStrictMode
```
Comment (zeripath) [inline modules/setting/session.go:64]:
```suggestion
		SessionConfig.SameSite = http.SameSiteLaxMode
```

File: modules/setting/session.go [hunk 8]
```diff
@@ -52,6 +56,15 @@ func newSessionService() {
 	SessionConfig.Gclifetime = sec.Key("GC_INTERVAL_TIME").MustInt64(86400)
 	SessionConfig.Maxlifetime = sec.Key("SESSION_LIFE_TIME").MustInt64(86400)
 	SessionConfig.Domain = sec.Key("DOMAIN").String()
+	samesiteString := sec.Key("SAME_SITE").In("strict", []string{"none", "lax", "strict"})
```
Comment (zeripath) [inline modules/setting/session.go]:
```suggestion
	samesiteString := sec.Key("SAME_SITE").In("lax", []string{"none", "lax", "strict"})
```

File: modules/setting/session.go [hunk 9]
```diff
@@ -31,10 +32,13 @@ var (
 		Secure bool
 		// Cookie domain name. Default is empty.
 		Domain string
+		// SameSite declares if your cookie should be restricted to a first-party or same-site context. Valid strings are "none", "lax", "strict". Default is "strict"
+		SameSite http.SameSite
 	}{
 		CookieName:  "i_like_gitea",
 		Gclifetime:  86400,
 		Maxlifetime: 86400,
+		SameSite:    http.SameSiteStrictMode,
```
Comment (zeripath) [inline modules/setting/session.go]:
```suggestion
		SameSite:    http.SameSiteLaxMode,
```

File: modules/setting/session.go [hunk 10]
```diff
@@ -31,10 +32,13 @@ var (
 		Secure bool
 		// Cookie domain name. Default is empty.
 		Domain string
+		// SameSite declares if your cookie should be restricted to a first-party or same-site context. Valid strings are "none", "lax", "strict". Default is "strict"
```
Comment (zeripath) [inline modules/setting/session.go]:
```suggestion
		// SameSite declares if your cookie should be restricted to a first-party or same-site context. Valid strings are "none", "lax", "strict". Default is "lax"
```

**Comment by CirnoT** [comment]
Certain cookies such as flash messages or transient session (2FA state, CAPTCHA or w/e) ones can be Strict as they refer to a single observed browsing session.

**Comment by zeripath** [comment]
OK well I think we can differentiate the few other cookies later. I think setting everything to lax for the moment is going to be correct enough.

**Pull request merged**